src/hooks.h

```diff
@@ -0,0 +1,85 @@
+struct HookSignaturesMatch<R(T1::*)(HookArgs...), R(T2::*)(HookArgs...)> {
+  static constexpr bool value = true;
+};
+
+// Equivalent to allow for const-hooks, e.g. bool getFooHook() const
+//
+template<typename R, typename T1, typename T2, typename... HookArgs>
+struct HookSignaturesMatch<R(T1::*)(HookArgs...) const, R(T2::*)(HookArgs...) const> {
+  static constexpr bool value = true;
+};
+
+// The above specialized HookSignaturesMatch plugin class is instanciated
+// based on two types of pointers to the plugin base class hook method and
+// the derived method. If both pointer types and thus the signatures equal,
+// the first or second specialization is instanciated,
+// leading to its compile time constant 'value' being defined as true.
+// Else the unspecialized version of the template class is instantiated,
+// causing 'value' to be defined as false.
+
+#define MATCH_HOOK_TYPE(HOOK_METHOD, PLUGIN)                                   \
+__NN__                                                                         \
+__NN__  {                                                                      \
+__NN__    /* Check it the two signatures equal. If not, the                    \
+__NN__    * user has implemented a method with the same name but a             \
+__NN__    * different (wrong) signature.                                       \
+__NN__    */                                                                   \
+__NN__    static_assert(HookSignaturesMatch<                                   \
+__NN__                       decltype(&KaleidoscopePlugin::HOOK_METHOD),       \
+__NN__                       decltype(&PLUGIN::HOOK_METHOD)                    \
+__NN__                  >::value,                                              \
+__NN__     "*****************************************************************" \
+__NN__     "******************** READ THIS CAREFULLY ! **********************" \
+__NN__     "**************************************************************** " \
+__NN__     "The signature of hook \"" #HOOK_METHOD                             \
+__NN__     "\" differs from the base class' hook signature. "                  \
+__NN__     "Please check all the hook methods implemented by your plugins "    \
+__NN__     "and compare them to the respective hook methods of "               \
+__NN__     "base class KaleidscopePlugin. Note: Two function signatures "      \
+__NN__     "equal if all argument types and the method's "                     \
+__NN__     "const qualifier equal. "                                           \
+__NN__     "*****************************************************************" \
+__NN__     "*****************************************************************" \
+__NN__     "*****************************************************************" \
+__NN__    );                                                                   \
+__NN__  }
```

Comment:
Missing newline leads to a compile error.

Reply:
Not sure what you mean. Is this about the newline at the end of the file? You are right, the standard says that there must be an empty line at the end of a header. Strange that my compiler did not report that. I will fix it with the next PR that improves the culprit plugin reporting.